modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoveryImpl.java

```diff
@@ -419,6 +447,32 @@ protected boolean checkAckTimeout(long ackTimeout) {
         return true;
     }
 
+    /** */
+    public void processMsg(TcpDiscoveryMetricsUpdateMessage msg, long tsNanos) {
+        for (Map.Entry<UUID, TcpDiscoveryMetricsUpdateMessage.MetricsSet> e : msg.metrics().entrySet()) {
+            UUID nodeId = e.getKey();
+
+            TcpDiscoveryMetricsUpdateMessage.MetricsSet metricsSet = e.getValue();
+
+            Map<Integer, CacheMetrics> cacheMetrics = msg.hasCacheMetrics(nodeId) ?
+                msg.cacheMetrics().get(nodeId) : Collections.emptyMap();
+
+            if (endTimeMetricsSizeProcessWait <= System.currentTimeMillis()
+                && cacheMetrics.size() >= METRICS_QNT_WARN)
+            {
+                log.warning("The Discovery message has metrics for " + cacheMetrics.size() +
+                    " caches.\nTo prevent Discovery blocking use -DIGNITE_DISCOVERY_DISABLE_CACHE=true option.");
+
+                endTimeMetricsSizeProcessWait = System.currentTimeMillis() + LOG_WARN_MSG_TIMEOUT;
```

Comment:
Please use org.apache.ignite.internal.util.IgniteUtils#currentTimeMillis ( U._currentTimeMillis_() ) instead of System.currentTimeMillis(). Also you can do just one call and save current timestamp to variable.